include/execution.hpp

```diff
@@ -2877,7 +2878,247 @@ namespace _P2300::execution {
+
+      public:
+        explicit __receiver(_SharedState& __shared_state) noexcept
+          : __shared_state_(__shared_state.__intrusive_from_this()) {
+        }
+
+        template <__one_of<set_value_t, set_error_t, set_stopped_t> _Tag, class... _As>
+        friend void tag_invoke(_Tag __tag, __receiver&& __self, _As&&... __as) noexcept {
+          _SharedState& __state = *__self.__shared_state_;
+
+          try {
+            using __tuple_t = __decayed_tuple<_Tag, _As...>;
+            __state.__data_.template emplace<__tuple_t>(__tag, (_As &&) __as...);
+          } catch (...) {
+            using __tuple_t = __decayed_tuple<set_error_t, std::exception_ptr>;
+            __state.__data_.template emplace<__tuple_t>(set_error, std::current_exception());
+          }
+
+          __state.__notify();
+          __self.__shared_state_.reset();
+        }
+
+        friend __env tag_invoke(get_env_t, const __receiver& __self) {
+          auto __stok = __self.__shared_state_->__stop_source_.get_token();
+          return make_env(with(get_stop_token, std::move(__stok)));
+        }
+      };
+
+    struct __operation_base {
+      using __notify_fn = void(__operation_base*) noexcept;
+      __notify_fn* __notify_{};
+    };
+
+    template <class _SenderId>
+      struct __sh_state : __enable_intrusive_from_this<__sh_state<_SenderId>> {
+        using _Sender = __t<_SenderId>;
+
+        template <class... _Ts>
+          using __bind_tuples =
+            __mbind_front_q<
+              __variant,
+              std::tuple<set_stopped_t>, // Initial state of the variant is set_stopped
+              std::tuple<set_error_t, std::exception_ptr>,
+              _Ts...>;
+
+        using __bound_values_t =
+          __value_types_of_t<
+            _Sender,
+            __env,
+            __mbind_front_q<__decayed_tuple, set_value_t>,
+            __q<__bind_tuples>>;
+
+        using __variant_t =
+          __error_types_of_t<
+            _Sender,
+            __env,
+            __transform<
+              __mbind_front_q<__decayed_tuple, set_error_t>,
+              __bound_values_t>>;
+
+        using __receiver_t = __receiver<__sh_state>;
+
+        __variant_t __data_;
+        in_place_stop_source __stop_source_{};
+
+        std::atomic<void*> __op_state1_;
+        connect_result_t<_Sender, __receiver_t> __op_state2_;
+
+        explicit __sh_state(_Sender& __sndr)
+          : __op_state1_{nullptr}
+          , __op_state2_(connect((_Sender&&) __sndr, __receiver_t{*this})) {
+          start(__op_state2_);
+        }
+
+        void __notify() noexcept {
+          void* __old{nullptr};
+          void* const __completion_state = static_cast<void*>(this);
+          bool const changed = __op_state1_.compare_exchange_weak(
+            __old, __completion_state,
+            std::memory_order_release,
+            std::memory_order_acquire);
+          if (changed && __old != nullptr) {
```

Comment:
I think this should be:
```suggestion
          void* const __completion_state = static_cast<void*>(this);
          void* const __old =
            __op_state1_.exchange(__completion_state, std::memory_order_acq_rel);
          if (__old != nullptr) {
```

@msimberg ?

Reply:
Yeah, that looks good to me!